solutions/src/main/java/ru/sberbank/school/task02/services/implementation/ExtendedFxConversionServiceImpl.java

```diff
@@ -0,0 +1,137 @@
+package ru.sberbank.school.task02.services.implementation;
+
+import ru.sberbank.school.task02.ExtendedFxConversionService;
+import ru.sberbank.school.task02.services.InternalQuotesService;
+import ru.sberbank.school.task02.util.Beneficiary;
+import ru.sberbank.school.task02.util.ClientOperation;
+import ru.sberbank.school.task02.util.Quote;
+import ru.sberbank.school.task02.util.Symbol;
+
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.Optional;
+
+import static ru.sberbank.school.task02.util.Beneficiary.BANK;
+import static ru.sberbank.school.task02.util.Beneficiary.CLIENT;
+import static ru.sberbank.school.task02.util.ClientOperation.BUY;
+import static ru.sberbank.school.task02.util.ClientOperation.SELL;
+
+/**
+ * Обратный валютный кальклятор: возвращает значение цены единицы котируемой валюты для заданного объема базовой.
+ * Для параметров SELL, USD/RUB, 1000, CLIENT, ситуация читается так:
+ * Клиент хочет продать долларов на 1000 рублей, сказать сколько будет стоить 1 доллар.
+ * Created by Gregory Melnikov at 01.05.2019
+ */
+public class ExtendedFxConversionServiceImpl extends FxConversionServiceImpl implements ExtendedFxConversionService {
+
+    private final InternalQuotesService internalQuotesService;
+    
+    ExtendedFxConversionServiceImpl(InternalQuotesService internalQuotesService) {
+        super(internalQuotesService);
+        this.internalQuotesService = internalQuotesService;
+    }
+
+    /**
+     * Возвращает значение цены единицы котируемой валюты для указанного объема
+     * только в том случае если есть полное совпадение.
+     *
+     * @param operation   вид операции
+     * @param symbol      Инструмент
+     * @param amount      Объем
+     * @param beneficiary В чью пользу осуществляется округление
+     * @return Цена для указанного объема
+     */
+    @Override
+    public Optional<BigDecimal> convertReversed(ClientOperation operation,
+                                                Symbol symbol,
+                                                BigDecimal amount,
+                                                Beneficiary beneficiary) {
+
+        List<Quote> quotes = internalQuotesService.getQuotes(symbol);
+
+        Quote targetQuote = findQuote(quotes, operation, amount, beneficiary);
+
+        BigDecimal quoteRate = operation == BUY ? targetQuote.getOffer() : targetQuote.getBid();
+
+        BigDecimal checkBaseValue = amount.divide(quoteRate, quoteRate.scale());
+
+        return checkBaseValue.compareTo(targetQuote.getVolumeSize()) == 0 ? Optional.ofNullable(quoteRate) : Optional.empty();
```

Comment:
Line is longer than 120 characters (found 126).

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a-1085548395>*